t/semaphore.t

```diff
@@ -0,0 +1,1431 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+use lib 'lib';
+use Test::Nginx::Socket::Lua;
+use Cwd qw(cwd);
+
+#worker_connections(10140);
+master_process_enabled(1);
+workers(1);
+log_level('warn');
+
+repeat_each(2);
+
+plan tests => repeat_each() * (blocks() * 3);
+
+my $pwd = cwd();
+
+no_long_string();
+#no_diff();
+
+#my $lua_default_lib = "/usr/local/openresty/lualib/?.lua";
+our $HttpConfig = <<_EOC_;
+    lua_package_path "$pwd/lib/?.lua;";
+_EOC_
+
+our $HttpConfigInitByLua= <<_EOC_;
+    lua_package_path "$pwd/lib/?.lua;";
+    init_by_lua '
+            require "resty.core.semaphore"
+            local sem  = ngx.semaphore.new(0)
+            local ok,err=sem:wait(1)
+            ngx.semaphore.err = err
+    ';
+_EOC_
+
+our $HttpConfigIntWorkerBy = <<_EOC_;
+    lua_package_path "$pwd/lib/?.lua;";
+    init_worker_by_lua '
+            require "resty.core.semaphore"
+            local sem  = ngx.semaphore.new(0)
+            local ok,err=sem:wait(1)
+            ngx.semaphore.test = err
+    ';
+
+_EOC_
+
+run_tests();
+
+__DATA__
+
+
+=== TEST 1: basic semaphore wait post
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        access_log off;
+        content_by_lua '
+        local res1, res2 = ngx.location.capture_multi{
+          { "/sub_sem_wait"},
+          { "/sub_sem_post"},
+        }
+        ngx.say(res1.status)
+        ngx.say(res1.body)
+        ngx.say(res2.status)
+        ngx.say(res2.body)
+        ';
+    }
+     location /sub_sem_wait {
+        proxy_pass $scheme://127.0.0.1:$server_port/sem_wait;
```

Comment:
@cuiweixie I know the default worker setting is 1. What I'm saying is that this test case is making things unnecessarily confusing and complex. Please keep in mind some advanced people check out test cases for more detailed documentation. We definitely should not encourage strange usage like this.

I think you can just use (0-delay) timers for this testing purpose where timer handlers have their own request contexts. It's also simpler than subrequests. Being simpler means being easier to debug things when tests fail. Even if you want to test subrequests in particular, you don't have to use `proxy_pass` to do HTTP communication in between.